### PR TITLE
validation support for pipeline parallelism [WIP]

### PR DESCRIPTION
**Pull request by wesleytruong**
With recent api change to pipeline schedule https://github.com/pytorch/pytorch/pull/157795, we can now schedule forward pass and calculate loss, allowing us to use validation and pp together.

To test correctness we train from a seed checkpoint with training.seed and training.determinism set with varying degrees of parallelism and different pipeline schedules to compare if loss remains the same:

| Parallelism | Loss |
| --- | --- |
| FSDP=2 | Screenshot 2025-07-29 at 5 12 49 PM |
| FSDP=2, TP=2, PP=2, PP_schedule="1F1B" |  Screenshot 2025-07-29 at 5 17 18 PM  |
| FSDP=2, PP=4, PP_schedule="1F1B" | Screenshot 2025-07-29 at 5 15 53 PM |
| FSDP=2, PP=4, PP_schedule="Interleaved1F1B" |Screenshot 2025-07-29 at 5 39 39 PM |
| FSDP=2, PP=4, PP_schedule="GPipe" | Screenshot 2025-07-29 at 5 49 36 PM
| FSDP=2, PP=4, PP_schedule="LoopedBFS" | Screenshot 2025-07-29 at 5 54 55 PM
| FSDP=2, PP=4, PP_schedule="InterleavedZeroBubble" | Screenshot 2025-07-30 at 2 30 53 PM

### CODE REVIEW

File: torchtitan/train.py [hunk 1]
```diff
@@ -341,6 +338,13 @@ def __init__(self, job_config: JobConfig):
                 validation_context=self.train_context,
                 maybe_enable_amp=self.maybe_enable_amp,
                 metrics_processor=self.metrics_processor,
+                pp_schedule=self.pp_schedule if parallel_dims.pp_enabled else None,
```
Comment (tianyu-l) [inline torchtitan/train.py]: maybe better to
```python
if parallel_dims.pp_enabled:
  pp_schedule, pp_has_first_stage, pp_has_last_stage = self.pp_schedule, self.pp_has_first_stage, self.pp_has_last_stage
else:
  pp_schedule, pp_has_first_stage, pp_has_last_stage = None, None, None
```
before this `build_validator_fn`

File: tests/integration_tests.py [hunk 2]
```diff
@@ -552,6 +552,18 @@ def build_test_list():
             "validation_fsdp_tp_cp",
             ngpu=8,
         ),
+        OverrideDefinitions(
+            [
+                [
+                    "--validation.enabled",
+                    "--validation.dataset c4_test",
+                    "--parallelism.pipeline_parallel_degree=2",
```
Comment (tianyu-l) [inline tests/integration_tests.py:548]: Let's merge this with the one above by replacing the dp_replicate, to save some CI resource.

Also please use a more sophisticated PP schedule, maybe `Interleaved1F1B`.

**Review by tianyu-l — APPROVED** [review]
LGTM!

**Review by H-Huang — APPROVED** [review]
Nice! https://github.com/pytorch/pytorch/pull/159475 will be landing for zero bubble soon
